File: upa-impl/src/main/csharp/Sources/Custom/Config/MakeSessionAwareMethodInterceptor.cs
using Net.Vpc.Upa.Impl.Util;

namespace Net.Vpc.Upa.Impl.Config
{


    /**
    * @author Taha BEN Salah
    * @creationdate 1/7/13 2:32 PM*/
    internal class MakeSessionAwareMethodInterceptor<T> {

        protected internal static System.Diagnostics.TraceSource log = new System.Diagnostics.TraceSource(typeof(Net.Vpc.Upa.Impl.Config.MakeSessionAwareMethodInterceptor).FullName);
        private readonly Net.Vpc.Upa.MethodFilter methodFilter;

        private readonly T instance;

        private Net.Vpc.Upa.Impl.Config.DefaultUPAContext defaultUPAContext;

        public MakeSessionAwareMethodInterceptor(Net.Vpc.Upa.Impl.Config.DefaultUPAContext defaultUPAContext, Net.Vpc.Upa.MethodFilter methodFilter, T instance) {
            this.defaultUPAContext = defaultUPAContext;
            this.methodFilter = methodFilter;
            this.instance = instance;
        }


        public virtual T MakeSessionAware() /* throws Net.Vpc.Upa.Exceptions.UPAException */
        {
            return instance.Proxy<T>(MyInterceptorMethod);
        }

        public object MyInterceptorMethod<T>(T@object, string methodName, object[] args, EmitProxyExecute<T> execute)
        {
            if (methodFilter == null /*|| methodFilter.Accept(method)*/) {
                System.Collections.Generic.IDictionary<string,object> properties = new System.Collections.Generic.Dictionary<string,object>();
                Net.Vpc.Upa.UPAContext context = Net.Vpc.Upa.UPA.GetContext();
                context.BeginInvocation(method, properties);
                object ret = null;
                System.Exception error = null;
                try {
                    ret = execute(@object, args);
                } catch (System.Exception e) {
                    error = e;
                }
                context.EndInvocation(error, properties);
                if(error!=null){
                    throw error;
                }
                return ret;
            } else {
                return execute(@object, args);
            }
        }
    }
}

File: upa-impl/src/main/java/net/vpc/upa/impl/config/MakeSessionAwareMethodInterceptor.java
package net.vpc.upa.impl.config;

import net.vpc.upa.MethodFilter;
import net.vpc.upa.UPA;
import net.vpc.upa.UPAContext;
import net.vpc.upa.impl.util.PlatformMethodProxy;
import net.vpc.upa.impl.util.PlatformMethodProxyEvent;

import java.util.HashMap;
import java.util.logging.Logger;

/**
 * @author Taha BEN SALAH <taha.bensalah@gmail.com>
 * @creationdate 1/7/13 2:32 PM
 */
public class MakeSessionAwareMethodInterceptor<T> implements PlatformMethodProxy<T> {

    private final MethodFilter methodFilter;
    private final T instance;
    private DefaultUPAContext defaultUPAContext;
    private static final Logger log = Logger.getLogger(MakeSessionAwareMethodInterceptor.class.getName());

    public MakeSessionAwareMethodInterceptor(DefaultUPAContext defaultUPAContext, MethodFilter methodFilter, T instance) {
        this.defaultUPAContext = defaultUPAContext;
        this.methodFilter = methodFilter;
        this.instance = instance;
    }

    @Override
    public Object intercept(PlatformMethodProxyEvent<T> event) throws Throwable {
        if (methodFilter == null || methodFilter.accept(event.getMethod())) {
            HashMap<String, Object> properties = new HashMap<String, Object>();
            UPAContext context = UPA.getContext();
            context.beginInvocation(event.getMethod(), properties);
            Object ret = null;
            Throwable error = null;
            try {
                ret = event.invokeBase(instance, event.getArguments());
            } catch (Throwable e) {
                error = e;
            }
            context.endInvocation(error, properties);
            if (error != null) {
                throw error;
            }
            return ret;
        } else {
            return event.invokeBase(instance, event.getArguments());
        }
    }
}
